library/ssl_cli.c

```diff
@@ -307,29 +307,35 @@ static int ssl_write_supported_elliptic_curves_ext( mbedtls_ssl_context *ssl,
 {
     unsigned char *p = buf;
     unsigned char *elliptic_curve_list = p + 6;
+    const size_t curve_list_size = (size_t) ( end - p - 6 );
     size_t elliptic_curve_len = 0;
     const mbedtls_ecp_curve_info *info;
-    const mbedtls_ecp_group_id *grp_id;
-
+    const uint16_t *tls_id;
     *olen = 0;
 
     MBEDTLS_SSL_DEBUG_MSG( 3,
         ( "client hello, adding supported_elliptic_curves extension" ) );
 
-    if( ssl->conf->curve_list == NULL )
+    if( mbedtls_get_supported_groups( ssl ) == NULL )
         return( MBEDTLS_ERR_SSL_BAD_CONFIG );
 
-    for( grp_id = ssl->conf->curve_list;
-         *grp_id != MBEDTLS_ECP_DP_NONE;
-         grp_id++ )
+    for( tls_id = mbedtls_get_supported_groups( ssl );
+         *tls_id != 0;
+         tls_id++ )
     {
-        info = mbedtls_ecp_curve_info_from_grp_id( *grp_id );
+        info = mbedtls_ecp_curve_info_from_tls_id( *tls_id );
         if( info == NULL )
         {
             MBEDTLS_SSL_DEBUG_MSG( 1,
                 ( "invalid curve in ssl configuration" ) );
             return( MBEDTLS_ERR_SSL_BAD_CONFIG );
         }
+
+        /* Check there is room for another curve */
+        if ( elliptic_curve_len + 1 >= curve_list_size )
+            return( MBEDTLS_ERR_SSL_BUFFER_TOO_SMALL );
+
+        MBEDTLS_PUT_UINT16_BE( *tls_id, elliptic_curve_list, elliptic_curve_len );
         elliptic_curve_len += 2;
```

Comment:
Am I blind, or is the increment of `elliptic_curve_list` missing?

Reply:
`elliptic_curve_list` points to the start of the curve_list, while `elliptic_curve_len` is an offset to the next two bytes to be filled with tls_id, and is incremented after each write to the list.

Reply:
Probably worth having a look at the 1.3-specific coding guidelines which I think suggest a pattern for that: https://github.com/ARMmbed/mbedtls/blob/development/docs/architecture/tls13-experimental.md#coding-rules-checklist-for-tls-13

(Note: I haven't actually looked at the code, just saw a notification for this comment and thought this might be relevant. Apologies if I guessed wrong and it's not.)